plugins/deisa/deisa.cxx

```diff
@@ -0,0 +1,228 @@
+				//
+			} else if (key == "logging" || key == "init_on") {
+				//
+			} else {
+				throw Config_error{key_tree, "Unknown key in Deisa file configuration: '{}'", key};
+			}
+		});
+
+
+		int mpi_size;
+		m_rank = Expression{Ref_r{ctx.desc("MPI_COMM_WORLD_rank").ref()}.scalar_value<long>()};
+		MPI_Comm comm = *static_cast<const MPI_Comm*>(Ref_r{ctx.desc("MPI_COMM_WORLD").ref()}.get());
+		MPI_Comm_size(comm, &mpi_size);
+		m_size = Expression{static_cast<long>(mpi_size)};
+
+
+		// plugin init
+		PC_tree_t init_tree = PC_get(conf, ".init_on");
+		if (!PC_status(init_tree)) {
+			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(); }, to_string(init_tree));
+		} else {
+			throw Config_error{conf, "Deisa plugin requires init_on key "};
+		}
+
+		// map_in
+		PC_tree_t map_tree = PC_get(conf, ".map_in");
+		if (!PC_status(map_tree)) {
+			each(map_tree, [&](PC_tree_t key_map, PC_tree_t value_map) {
+				ctx.callbacks().add_data_callback(
+					[&, deisa_array_name = to_string(value_map)](const std::string&, const Ref& data_ref) {
+						try {
+							// start a python context and call bridge.publish_data(...)
+							pydict pyscope = pymod::import("__main__").attr("__dict__");
+							pyscope[deisa_array_name.c_str()] = to_python(data_ref);
+							pyscope["time_step"] = m_time_step.to_long(ctx);
+							pyscope["name"] = deisa_array_name.c_str();
+
+#ifdef NDEBUG
+							py::exec("bridge.publish_data(" + deisa_array_name + ", name, time_step, debug=False)", pyscope);
+#else
+							py::exec("bridge.publish_data(" + deisa_array_name + ", name, time_step, debug=True)", pyscope);
+#endif
+							pyscope[deisa_array_name.c_str()] = NULL;
+						} catch (const std::exception& e) {
+							ctx.logger().error("While publishing data, caught exception {}", e.what());
+							throw Plugin_error("While publishing data. Caught exception: {}", std::string(e.what()));
+						} catch (...) {
+							ctx.logger().error("While publishing data.");
+							throw Plugin_error("While publishing data.");
+						}
+					},
+					to_string(key_map)
+				);
+			});
+		}
+	}
+
+	~deisa_plugin() noexcept override
+	{
+		try {
+			if (m_interpreter_initialized_in_plugin) {
+				py::exec(R"(
+				if bridge:
+				  bridge.release()
+                		)"); // call bridge release() so that we can clear things before destructor is called (if it is ever called !).
+				py::finalize_interpreter();
+			}
+		} catch (const std::exception& e) {
+			context().logger().error("Exception in destructor, caught exception {}", e.what());
+		} catch (...) {
+			context().logger().error("Exception in destructor");
+		}
+	}
+
+private:
+	/**
+	 * Check that the PDI plugin is compatible with Deisa's python Bridge (i.e, that the python API is what it should be).
+	 * The plugin is compatible if PYTHON_LIBRARY_COMPATIBILITY == deisa.__version__.__version__ (python)
+	 */
+	static void check_compatibility()
+	{
+		pymod deisa = pymod::import("deisa.__version__");
```

Comment:
I'm not going to, because it creates a segfault when calling `py::finalize_interpreter();` and this is only called once. If a scenario occurs where the module is used more than once, I'll consider making this a class variable.

Reply:
yes, it should be set to Null in the dtor before finalizing, but indeed, now that you store m_bridge, it makes no sense to store the module.